composer/trainer/trainer.py

```diff
@@ -996,7 +996,6 @@ def __init__(
             # Always ignore the run_name in the checkpoint so it is consistent with what was used for Event.INIT.
             self.state.run_name = run_name
             log.info(f'Setting seed to {self.state.seed}')
-            reproducibility.seed_all(self.state.seed)
```

Comment:
Hmm, curious why was this line removed? I believe it is needed for the checkpointing tests...

Reply:
Random number generator state is restored from the checkpoint. We don't need to seed them again. If we do, resumption from a checkpoint won't match as the generated random numbers (If an algorithm uses them) would be different.

Reply:
Removed the log statement and comment as well.

Reply:
Ah, I needed a night's sleep to remember the issue here:

The seed must be restored here, as the dataloader initializes its sampler on the first spin, not during Event.INIT. Without this, then the dataloader sampler will be initialized with a different random value than the seed in the checkpoint.

Reply:
Maybe we should add a comment there to record this for posterity :)

Reply:
Agreed with hanlin this needs to be better documented haha :)

Following the code path in `trainer.py`, here's what happens w.r.t to seed and RNG state.

1. Run `Event.INIT` -- nothing should depend on the seed / rng here (line 874)
1. Load checkpoint (line 987)
2. Restore the seed (line 999)
3. Run `Event.FIT_START` (line 1420). Nothing should use the seed here.
4. Spin the dataloaders (line 1428). This will cause the dataloaders to read the RNG from the seed
5. Before the first batch that will be trained after the checkpoint, restore the RNG state, so it is accurate for events that will be used by the algorithms (line 1432 and line 1464, depending on whether it is a batch or epoch checkpoint)

So, in effect, the `state.seed` is used for the dataloaders, and `state.rng` is used for the algorithms